File: pset1/credit.c
#include <stdio.h>
#include <math.h>
#include <cs50.h>

int main(void)
{
    long long number;
    //get input
    do
    {
        number = get_long("Number: ");
    }
    while (number < 0 || number > 9999999999999999);

    //multiply every other digit by 2, starting with second last digit
    //add digits of products together
    //add digits of products together
    //add sum to sum of digits not multiplied by 2
    //if total's digit = 0, card is legit
    long long arr[16];

    //luhns algorithm
    for (int i = 0; i < 16; i++)
    {
        long long x = pow(10, i + 1);
        arr[i] = number % x;
    }

    for (int i = 15; i >= 0; i--)
    {
        long long x = pow(10, i);
        if (i != 0)
        {
            arr[i] = arr[i] - arr[i - 1];
        }
        arr[i] = arr[i] / x;
    }

    int check = 0;
    int tmpDigit = 0;
    for (int i = 1; i < 16; i += 2)
    {
        tmpDigit = arr[i] * 2;
        if (tmpDigit >= 10)
        {
            tmpDigit -= 10;
            check += 1;
        }
        check += tmpDigit;
    }

    for (int i = 0; i < 16; i += 2)
    {
        check += arr[i];
    }

    //printf("%i\n", check);
    //check company of card
    if (check % 10 == 0 && number >= 1000000000000)
    {
        if (arr[15] == 4 ||
            (arr[15] == 0 && arr[14] == 4) ||
            (arr[15] == 0 && arr[14] == 0 && arr[13] == 4) ||
            (arr[15] == 0 && arr[14] == 0 && arr[13] == 0 && arr[12] == 4))
        {
            printf("VISA\n");
        }
        else if (arr[15] == 0 && arr[14] == 3 &&
                 (arr[13] == 4 || arr[13] == 7))
        {
            printf("AMEX\n");
        }
        else if (arr[15] == 5 && (arr[14] >= 1 && arr[14] <= 5))
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

File: 2019_psets/pset6/credit/credit.py
#!/usr/bin/env python3

while True:
    try:
        # Get card number in string format
        card_number = int(input("Number: "))
    except ValueError:
        continue

    card_string = str(card_number)
    # Check Luhn's Algorithm
    luhn_sum = 0
    temp_digit = 0
    # Flip string card number around
    luhn_string = card_string[::-1]

    for i in range(len(luhn_string)):
        # For every other digit from 2nd to last digit of card_string
        if i % 2 == 1:
            temp_digit = 2 * int(luhn_string[i])
            luhn_sum += (temp_digit % 10) + (temp_digit // 10)

        else:
            luhn_sum += int(luhn_string[i])

    # Check that last digit is 0
    if luhn_sum % 10 == 0:
        # Check if AMEX: 15 digits, start with 34 or 37
        if len(card_string) == 15 and card_string[0] == '3' and (card_string[1] == '4' or card_string[1] == '7'):
            print('AMEX')
            break
        # Check if Mastercard: 16 digits, start with 51, 52, 53, 54, 55
        elif len(card_string) == 16 and card_string[0] == '5' and int(card_string[1]) in range(1, 6):
            print('MASTERCARD')
            break
        # Check if VISA: 13 and 16 digits
        elif (len(card_string) == 13 or len(card_string) == 16) and card_string[0] == '4':
            print('VISA')
            break

        else:
            print('INVALID')
            break
    # If not amex, mastercard or visa
    else:
        print('INVALID')
        break


# Output AMEX, MASTERCARD, VISA OR INVALID